### PR TITLE
Add models.py

### PR DESCRIPTION
**Pull request by arbennett**
This adds `models.py`, which is the module that contains all of the functionality for training a machine learning method. Generally, it provides several things:
 - Preprocessing steps (data transformation/normalization, partitioning test/train data, creating lookback datasets for LSTMs)
 - Model creation (functions for creating LSTMs and BidirectionalLSTMs currently)
 - Prediction (Taking a trained model, and input data, and running a full workflow on it)

### CODE REVIEW

File: bmorph/core/models.py [hunk 1]
```diff
@@ -0,0 +1,361 @@
...
+            ('log1p', FunctionTransformer(func=np.log1p, inverse_func=np.expm1,
+                                          validate=False)),
+            ('normalize', MinMaxScaler())
+            ]),
+        'precipitation': Pipeline([
+            ('cbrt', FunctionTransformer(func=np.cbrt, inverse_func=cube,
+                                         validate=False)),
+            ('normalize', MinMaxScaler())
+            ]),
+        'elevation': FunctionTransformer(elevation_scaler),
```
Comment (YifanCheng) [inline bmorph/core/models.py]: Why do you normalize elevation differently than other variables?

File: bmorph/core/models.py [hunk 2]
```diff
@@ -0,0 +1,361 @@
...
+    a machine learning model. Note that this does not actually do
+    any data transformations, but merely fits the transformers
+    and returns them for later application. Any variables not in
+    the standard set provided here will simply be standardized
+
+    Parameters
+    ----------
+    df:
+        Dataframe with complete set of all data for training
+        This should have a MultiIndex with levels ('site', 'time')
```
Comment (YifanCheng) [inline bmorph/core/models.py]: Please clarify whether the data will be standardized across multiple sites or each site will be standardized individually.

File: bmorph/core/models.py [hunk 3]
```diff
@@ -0,0 +1,361 @@
...
+        List of sites to partition
+    train_frac:
+        Fraction of sites to place into the training partition
+
+    Returns
+    -------
+    train_sites, test_sites
+    """
+    n_train = int(train_frac * len(sites))
+    shuffled = copy.copy(sites)
```
Comment (YifanCheng) [inline bmorph/core/models.py]: Why not directly shuffle `sites`?

File: bmorph/core/models.py [hunk 4]
```diff
@@ -0,0 +1,361 @@
...
+    -------
+    train_df, test_df
+    """
+    sites = np.unique(df.index.get_level_values('site'))
+    train_sites, test_sites = split_train_test_sites(sites, train_frac)
+    return df.loc[train_sites], df.loc[test_sites]
+
+
+def make_lookback(df: pd.DataFrame, lookback: int=7) -> pd.DataFrame:
+    """
```
Comment (YifanCheng) [inline bmorph/core/models.py]: add a little description about this function

File: bmorph/core/models.py [hunk 5]
```diff
@@ -0,0 +1,361 @@
...
+    in_features = list(lookback_ds.features.values)
+    in_features.remove(target_feature)
+    lstm_features = lookback_ds.sel(features=in_features)
+    lstm_target = (lookback_ds.sel(features=target_feature)
+                              .isel(lookback=-1))
+    return lstm_features, lstm_target
+
+
+def make_callbacks(name):
+    from tensorflow.keras.callbacks import EarlyStopping
```
Comment (YifanCheng) [inline bmorph/core/models.py]: add a description of this function

File: bmorph/core/models.py (hunk 2 above)
Comment (arbennett) [reply bmorph/core/models.py]: Data must be transformed with respect to all training data, so multiple sites.

File: bmorph/core/models.py (hunk 1 above)
Comment (arbennett) [reply bmorph/core/models.py]: Since we a priori know the range of elevations on Earth, we can just use them as the minimum and maximum values. That means that our transformer is not sensitive to biases in site selection.

File: bmorph/core/models.py (hunk 3 above)
Comment (arbennett) [reply bmorph/core/models.py]: Calling `random.shuffle` mutates the object in place, which would change the `sites` list. This introduces some pretty unintuitive behavior. So, we make a copy and shuffle that so that we are not messing with a user's variable.
